src/Common/FileSegment.cpp

```diff
@@ -684,29 +664,39 @@ bool FileSegment::hasFinalizedState() const
         || download_state == State::SKIP_CACHE;
 }
 
-void FileSegment::detach(std::lock_guard<std::mutex> & cache_lock, std::lock_guard<std::mutex> & segment_lock)
+bool FileSegment::isForcefullyDetached() const
 {
-    if (detached)
+    std::lock_guard detach_lock(detach_mutex);
+    return is_forcefully_detached;
+}
+
+void FileSegment::detach(
+    bool forced_detach,
+    std::lock_guard<std::mutex> & /* cache_lock */,
+    std::lock_guard<std::mutex> & /* detach_lock */,
+    std::lock_guard<std::mutex> & segment_lock)
+{
+    if (is_detached)
         return;
 
     markAsDetached(segment_lock);
```

Comment:
Is it possible to get rid of this function?

Reply:
No, it is used many times in FileCache.cpp, it sets `is_detached=true` and updated ProfileEvents, if there was no profile events update or if this method needed in no more than 3 places (now about 6), then it could be removed